src/debug_info.rs

```diff
@@ -1,30 +1,1221 @@
+            LLVMDIBuilderInsertDeclareAtEnd(
+                self.builder,
+                storage.as_value_ref(),
+                var_info
+                    .map(|v| v.metadata_ref)
+                    .unwrap_or(std::ptr::null_mut()),
+                expr.unwrap_or(self.create_expression(vec![])).metadata_ref,
+                debug_loc.metadata_ref,
+                block.basic_block,
+            )
+        };
+        InstructionValue::new(value_ref)
+    }
+
+    /// Create an expression
+    pub fn create_expression(&self, mut address_operations: Vec<i64>) -> DIExpression<'ctx> {
+        let metadata_ref = unsafe {
+            LLVMDIBuilderCreateExpression(
+                self.builder,
+                address_operations.as_mut_ptr(),
+                address_operations.len(),
+            )
+        };
+        DIExpression {
+            metadata_ref,
+            _marker: PhantomData,
+        }
+    }
 
-impl DebugInfoBuilder {
-    #[cfg(feature = "experimental")]
-    pub(crate) fn new(dib: LLVMDIBuilderRef) -> Self {
-        debug_assert!(!dib.is_null());
+    /// Insert a new llvm.dbg.value intrinsic call before an instruction.
+    pub fn insert_dbg_value_before(
+        &self,
+        value: BasicValueEnum<'ctx>,
+        var_info: DILocalVariable<'ctx>,
+        expr: Option<DIExpression<'ctx>>,
+        debug_loc: DILocation<'ctx>,
+        instruction: InstructionValue<'ctx>,
+    ) -> InstructionValue<'ctx> {
+        let value_ref = unsafe {
+            LLVMDIBuilderInsertDbgValueBefore(
+                self.builder,
+                value.as_value_ref(),
+                var_info.metadata_ref,
+                expr.unwrap_or(self.create_expression(vec![])).metadata_ref,
+                debug_loc.metadata_ref,
+                instruction.as_value_ref(),
+            )
+        };
+        InstructionValue::new(value_ref)
+    }
 
-        DebugInfoBuilder(dib)
+    /// Construct a placeholders derived type to be used when building debug info with circular references.
+    ///
+    /// All placeholders must be replaced before calling finalize() otherwise we will panic!().
+    pub fn create_placeholder_derived_type(&mut self, context: &Context) -> DIDerivedType<'ctx> {
+        let metadata_ref = unsafe { LLVMTemporaryMDNode(context.context, std::ptr::null_mut(), 0) };
+        self.placeholders.push(metadata_ref);
+        DIDerivedType {
+            metadata_ref,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Deletes a placeholder, replacing all uses of it with another derived type.
+    pub fn replace_placeholder_derived_type(
+        &mut self,
+        placeholder: DIDerivedType<'ctx>,
+        other: DIDerivedType<'ctx>,
+    ) {
+        // `placeholder` must be in the placeholder list.
+        // `other` may or may not be a placeholder.
+        let index = self
+            .placeholders
+            .iter()
+            .position(|ph| *ph == placeholder.metadata_ref)
```

Comment:
Does this intentionally stop at the first one? Can there be multiple?

Reply:
It was stopping at the first one and they needed to be unique. Regardless, this entire technique doesn't work (DebugInfoBuilder is Copy and then the `placeholders` member gets out of sync) so this has been deleted.